spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/kafka/KafkaAutoConfiguration.java

```diff
@@ -137,4 +146,22 @@ public KafkaAdmin kafkaAdmin() {
 		return kafkaAdmin;
 	}
 
+	@Bean
+	@ConditionalOnProperty(name = "spring.kafka.retry.topic.enabled")
+	public RetryTopicConfiguration kafkaRetryTopicConfiguration(KafkaOperations<Object, Object> kafkaOperations) {
+		KafkaProperties.Retry.Topic retryTopic = this.properties.getRetry().getTopic();
+		return RetryTopicConfigurationBuilder.newInstance().maxAttempts(retryTopic.getAttempts())
+				.customBackoff(getBackOffPolicy(retryTopic)).useSingleTopicForFixedDelays()
+				.suffixTopicsWithIndexValues().doNotAutoCreateRetryTopics().create(kafkaOperations);
+	}
+
+	private SleepingBackOffPolicy<?> getBackOffPolicy(Topic retryTopic) {
+		BackOffPolicy policy = BackOffPolicyBuilder.newBuilder().delay(retryTopic.getDelayMillis())
+				.maxDelay(retryTopic.getMaxDelayMillis()).multiplier(retryTopic.getMultiplier())
+				.random(retryTopic.isRandomBackOff()).build();
+		Assert.isInstanceOf(SleepingBackOffPolicy.class, policy,
+				() -> "BackOffPolicy must be an instance of SleepingBackOffPolicy. Provided: " + policy);
```

Comment:
I don't think it is OK to say this: the maximum what end-user can provide are exactly those props you use for `BackOffPolicyBuilder`.
So, we might need to ensure that only `SleepingBackOffPolicy` is created. Perhaps just ignore some props which are not sufficient...

Reply:
Thanks for looking into this @artembilan. Currently, the `BackOffPolicyBuilder` only returns instances of `SleepingBackOffPolicy`, but I didn't want to add this constraint to the builder's method signature.

So this assertion here is just for the case that, if we in the future update the `Builder` to for example return a `NoBackOffPolicy` if delay is set to `0`, we don't get a class cast exception, but get this error instead.

But I agree that this is not a really useful message to the user. Maybe we should just blindly cast? Not sure what a useful message for the users should be in this case, since it'd not really be their fault.

Reply:
If we don't support a `delay == 0`, then it has to be rejected as a configuration properties validation, before calling this builder.
What I mean, if we don't support anything else in the builder, then we need to ensure the props provided for it are in the valid range.
The message you say now is fully misleading: end-user does not create instances of this, so the message (if any), must really point what is wrong with end-user config.

Reply:
Well, currently if the provided delay < 1 it is set to 1 in the `FixedBackOffPolicy`, and most or all other `BackOffPolicy` implementations follow that pattern. In `RetryTopic`'s annotation processing and builder I chose to mimic `Spring Retry` behavior as much as possible, since we relied on the same `@BackOff` annotation.

I've been meaning to ask if maybe we should change this in `Spring Retry` to return the `NoBackOffPolicy` instead if `delay=0` is provided, but didn't want to mix things up. Also, that seems like a breaking change, so it should be a good fit only for `SR` 2.0.

Maybe it'd be worth it to validate user input here to make sure the inputs are > or >=  0. Although it might be confusing having different behaviors for auto configuration and the `@RetryableTopic` annotation and builder.

Reply:
So, why then does `RetryTopicConfigurationBuilder` support only a `SleepingBackOffPolicy`?
Why it cannot be that super `BackOffPolicy`?

Reply:
That's because the way we [generate the delay values](https://github.com/spring-projects/spring-kafka/blob/8ccdbd7a8a54815382645a8e9dd30cd82c0334dc/spring-kafka/src/main/java/org/springframework/kafka/retrytopic/BackOffValuesGenerator.java#L78-L92
) is we inject a `Sleeper` implementation in the `BackOffPolicy` that calls `backOff` and stores the values, so we need that any custom policy the user provides respect the `SleepingBackOffPolicy` contract.

The builder offers a `noBackOff()` method if that's what the user desires, and if that's the case we don't need the `Sleeper` to fetch the values.

Reply:
I see there `public RetryTopicConfigurationBuilder noBackoff() {`, so perhaps with some props config we can fall back to this instead of `customBackoff()` then?

Reply:
You mean, if users set `delay=0` we could set `noBackOff()` for them there? That's an interesting thought. My concern would be having different behaviors in auto configuration and annotation processing - would that be confusing for users? I guess since these are kind of corner cases, maybe we can depart from that and also add properties validation.

WDYT?

Reply:
Yes, that would be reasonable.

Thanks

Reply:
Sure, I’ll add that then. Thanks a lot for your input!